### PR TITLE
disallow multidimensional slices and arrays

### PR DESCRIPTION
**Pull request by tessr**
Stop allowing multidimensional slices and arrays, in order to increase proto3 compatibility. Addresses #272.

This is my first PR, and a little bit of a shot in the dark, so please let me know if I need to restructure anything!

### CODE REVIEW

**Comment by tessr** [comment]
Ah, just remembered this needs to permit multidimensional arrays of bytes. Will add

**Comment by tessr** [comment]
This now permits byte arrays and slices at any depth, and is ready for review!

**Review by tac0turtle — APPROVED** [review]
utACK ⚡️

**Comment by zmanian** [comment]
Manually checked both Tendermint and the CosmosSdk and `[][]byte` is the multidimensional struct we are actively serializing.

Actually trying to integrate this branch into tendermint and the sdk got me into a dependency resolution snowball.

**Review by zmanian — APPROVED** [review]

**Comment by tac0turtle** [comment]
oooh yea, you have to update roughly every repo of ours to make it work. We can merge this into master and then update ismail's branch with it or merge it into ismail's branch, which is preferred? @zmanian  then I can run through and update all the existing branches I created in iavl, tendermint and cosmos-sdk to test it with the simulator,

**Comment by tessr** [comment]
Good question, I'm not sure if it would be better to merge this into master or into Ismail's feature branch. Which is more copacetic with our normal git flow? :) 

Fwiw, this change could stand on its own, if we're trying to avoid bloating Ismail's PR.

**Comment by tac0turtle** [comment]
This repo and Iavl don't really follow our normal gitflow. They are outliers in a way, this repo won't receive a release till ismail's pr is complete so it may make more sense to merge to master then back update ismail's. Will do merge this into master then go about testing in the sdk again.

File: binary-encode.go [hunk 1]
```diff
@@ -66,13 +66,45 @@ func (cdc *Codec) encodeReflectBinary(w io.Writer, info *TypeInfo, rv reflect.Va
...
+				}
+			}
 		} else {
 			err = cdc.encodeReflectBinaryList(w, info, rv, fopts, bare)
 		}
 
 	case reflect.Slice:
 		if info.Type.Elem().Kind() == reflect.Uint8 {
 			err = cdc.encodeReflectBinaryByteSlice(w, info, rv, fopts)
+		} else if info.Type.Elem().Kind() == reflect.Slice || info.Type.Elem().Kind() == reflect.Array {
```
Comment (ebuchman) [inline binary-encode.go:93]: Can't we deduplicate these?

File: tests/proto3/proto3_compat_test.go [hunk 2]
```diff
@@ -108,6 +108,49 @@ func TestFixedU64Roundtrip(t *testing.T) {
 	assert.EqualValues(t, p3ToAm.Int64, amToP3.Int64)
 }
 
+func TestMultidimensionalSlices(t *testing.T) {
+	s := [][]int8{
+		[]int8{1, 2},
+		[]int8{3, 4, 5}}
+
+	_, err := cdc.MarshalBinaryBare(s)
+	assert.Error(t, err, "expected error: multidimensional slices are not allowed")
```
Comment (ebuchman) [inline tests/proto3/proto3_compat_test.go:117]: What about a test with more than 2 dimensions? And what about tests that mix arrays and slices (!)?

**Pull request merged**